### DISCUSSION TITLE
Custom `__str__` vs `__repr__`

### DISCUSSION THREAD

**Discussion by ajwood** [Textualize/rich#1847] [Q&A] [answered]
I'm trying to add some pretty colors to the console representation of a custom object. The [Console Prorotol](https://rich.readthedocs.io/en/stable/protocol.html) docs describe how to do exactly that, by defining a `__rich__` method. This method seems to override the built-in `__str__` (although I find the docs actually suggest it would be `__repr__`). This is great, so when I `rich.print(MyObj(123))`, its `__rich__` method kicks in and everything is beautiful. However with `rich.print([MyObj(123), MyObj(456)])`, it doesn't use the custom rendering.

I found the [Rich Repr Protocol](https://rich.readthedocs.io/en/stable/pretty.html#rich-repr-protocol), but that seems to solve a different problem.

Is it possible to do what I'm attempting, implement custom methods to replace both `__str__` and `__repr__` for the purposes of custom `rich.print` output?

**Comment by willmcgugan** [comment, marked as answer]
What would you expect printing a list of `MyObj` to do?

When you print a container like a list, you are invoke pretty printing, which is designed for displaying data structures.

**Reply by ajwood** [reply]
> What would you expect printing a list of MyObj to do?

Not sure exactly, but what rich does by default seems to make sense. IIUC, it applies some built-in logic to parse the `__repr_` string, and does its best to make it pretty.

The part that was surprising me is that the `__rich__` hook exists to define custom formatting for when python goes for its `__str__` representation, but there doesn't seem to be an equivalent for `__repr__` (I thought  `__rich_repr__` would be the answer, but that doesn't appear to support custom coloring). So I was expecting to find a way to have full control of how the custom object gets printed, regardless of whether python wants to render it with `__str__` or `__repr__`.

To quickly describe my current scenario, I have a filesystem tree that follows a structured layout. For an entry in this tree e.g. `aaa/bbb/ccc/ddd:1.2.3:4567`, trying to make it easy to look at, where each item in the colon-separated component comes out a different colour, even in the case where I'm printing a list of these path items.

**Reply by ajwood** [reply]
> IIUC, it applies some built-in logic to parse the __repr_ string, and does its best to make it pretty.

Oh I see where I went wrong there, the built-in parsing is on _list_'s `__str__` output, which happened to have called MyObj's `__repr__`. But by the time `rich.print` comes into the picture, it's too late?

**Reply by ajwood** [reply]
FWIW here's a working example of what I'm trying:
```python
import itertools

from rich import print


class Foo:
    def __init__(self, tag: str):
        """A toy class with a colon-separated tag"""
        self.tag = tag

    def __rich__(self) -> str:
        # alternate tag component colours
        colors = itertools.cycle(("red", "green"))
        x = "[blue]:[blue/]".join(
            "[{color}]{t}[{color}/]".format(t=t, color=next(colors))
            for t in self.tag.split(":")
        )
        return x

    def __rich_repr__(self) -> str:
        yield self.tag


f = Foo("aaa:bbb:1.2.3:4567")
print(f)  # renders with alternating colours
print([f, f])  # can't get `f` rendered with custom format logic
```

**Reply by willmcgugan** [reply]
I see. No, I'm afraid you're asking too much from Rich there. When you are printing a dictionary you are essentially asking for a repr of it and its contents. It won't do any of the fancy formatting that you can do via `__rich__`.

By the looks of it, what you want is *highlighting*. If you are familiar with regexes, you can create a [highlighter](https://rich.readthedocs.io/en/latest/highlighting.html) to automatically colorize certain patterns.

**Reply by ajwood** [reply]
Ohh indeed that _highlighting_ looks very promising, I'll check it out. Thanks for taking the time to help!
